### ISSUE TITLE
DocumentMatch.Sort does not agree with SearchBefore/SearchAfter

### ISSUE THREAD

**Issue by asuffield** [blevesearch/bleve#1610]
https://github.com/blevesearch/bleve/commit/60750c16140fe90492143f0eb9f39560e75f6124 is the change at the root of this problem. https://github.com/blevesearch/bleve/issues/1583 describes expected usage.

In the (default) case of sorting by score, DocumentMatch.Sort does not contain the score, because of the change linked above.

We probably want a different function on either DocumentMatch or SearchSort which returns the actual sort value.

**Comment by mschoch** [comment]
@abhinavdangeti pinging you on this, as you're investigating a different issue with scoring right now.

**Comment by abhinavdangeti** [comment]
@asuffield Firstly, when you sort by score (default), the DocumentMatch.Sort will just carry "_score" to indicate what was used for ordering the results. The DocumentMatch should have a separate field "score" indicating the actual score (numeric) of the document.

Note that the DocumentMatch.Sort is an array of strings. The encoded version (bytes) of the "_score" of every document is not the easiest to decode. If you try sorting your documents over a numeric field - you'll see what I mean.
To view the actual content of these fields, I'd recommend "store"ing the fields while indexing and requesting the particular field by specifying it within the "fields" section of the SearchRequest, to view the actual content as is within DocumentMatch.Fields.

Now, when you use SearchAfter/SearchBefore, the expectation is - **the length of the array for SearchAfter/SearchBefore should match with the sort array within the SearchRequest**.
If you're running SearchAfter/SearchBefore over a document ID, then the expectation is to use "_id" as the sort key to search and order your results properly. You will see the "sort" key appear as expected within DocumentMatch.Sort for all but _score, which should be available within DocumentMatch.Score.

**Comment by mschoch** [comment]
@abhinavdangeti I agree with everything you said, but I think now that we both misunderstood the issue.

My original understanding was that the issue being pointed out was that when we do a multi-search, we need to make sure that the sort we use compares the *actual score* at the appropriate place, and doesn't incorrectly compare the strings "_score" from one hit with "_score" from another hit.  I didn't ever confirm this, I just thought that you were already looking at an issue in multi-search result ordering.

However, I now think that @asuffield was simply asking for a helper method, on the DocumentMatch, which could compute the *actual* SortValue one would need, if you wanted to pass it as an argument the search_before or search_after.  This helper function would simply replace "_score" (and any other such cases) with the actual byte encoding of the value that is necessary.

So, I'd like to review how multi-search tries to do this correctly, verify it and go from there...

**Comment by asuffield** [comment]
> However, I now think that @asuffield was simply asking for a helper method, on the DocumentMatch, which could compute the _actual_ SortValue one would need, if you wanted to pass it as an argument the search_before or search_after. This helper function would simply replace "_score" (and any other such cases) with the actual byte encoding of the value that is necessary.

Correct - this is easy to do manually when it's {"_score"}, but remarkably difficult in any other cases, making it challenging to use SearchAfter with non-trivial sort ordering.

**Comment by mschoch** [comment]
@asuffield ok, so I'm confused again because I'm not sure what other cases there are, or why they are "remarkably difficult"?  When I said "and any other such cases" I was being careful because I thought possibly we also need a replace for "_id" as well.  But that is even simpler than the score because it requires no special encoding.  So what other cases are there?

**Comment by asuffield** [comment]
If the sort order is something like {field foo, score} and was parsed from json instead of written into the code, then doing this from the available api would end up looking something like:

```
value := make([]string, len(req.Sort))
for i, sort := range req.Sort {
  switch sort.(type) {
  case *search.SortScore:
    value[i] = strconv.FormatFloat(doc.Score, 'e', -1, 64)
  default:
    value[i] = doc.Sort[i]
  }
}
```

which knows way too much about the implementation details of github.com/blevesearch/bleve/v2/search and is challenging to figure out how to write correctly (I think the above is correct for the latest release, but I'm not sure).

This really wants to live in the search.SearchSort interface.
